modelstore/storage/gcloud.py

```diff
@@ -61,44 +63,68 @@ def __init__(
-            if self.__client is None:
-                self.__client = storage.Client(self.project_name)
+
+        if self.__client is not None:
             return self.__client
-        except DefaultCredentialsError:
-            try:
-                # Try to authenticate, if in a Colab notebook
-                # pylint: disable=no-name-in-module,import-error,import-outside-toplevel
-                from google.colab import auth
 
-                auth.authenticate_user()
+        if self.project_name is not None:
+            try:
                 self.__client = storage.Client(self.project_name)
-                return self.__client
-            except ModuleNotFoundError:
-                pass
-            logger.error(
-                "Missing credentials: https://cloud.google.com/docs/authentication/getting-started#command-line"  # noqa
-            )
-            raise
+            except DefaultCredentialsError:
+                try:
+                    # Try to authenticate, if in a Colab notebook
+                    # pylint: disable=no-name-in-module,import-error,import-outside-toplevel
+                    from google.colab import auth
+
+                    auth.authenticate_user()
+                    self.__client = storage.Client(self.project_name)
+                    return self.__client
+                except ModuleNotFoundError:
+                    logger.warning(
+                        "Missing credentials: https://cloud.google.com/docs/authentication/getting-started#command-line"  # noqa
+                    )
+                    warnings.warn(
+                        "No credentials given, falling back to anonymous access."
+                    )
+
+        self.__client = storage.Client.create_anonymous_client()
+        return self.__client
 
     def validate(self) -> bool:
         """Runs any required validation steps - e.g.,
         checking that a cloud bucket exists"""
         logger.debug("Querying for buckets with prefix=%s...", self.bucket_name)
-        for bucket in list(self.client.list_buckets(prefix=self.bucket_name)):
-            if bucket.name == self.bucket_name:
+        if self.client.project is not None:
+            for bucket in list(self.client.list_buckets(prefix=self.bucket_name)):
+                if bucket.name == self.bucket_name:
+                    return True
+            return False
+        else:
+            bucket = self.client.bucket(bucket_name=self.bucket_name)
+            try:
+                _ = list(bucket.list_blobs())
+            except (NotFound, BadRequest):
+                logger.error(
+                    f"Bucket '{self.bucket_name}' does not exist or is not accessible for anonymous clients."
+                )
+                return False
+            else:
                 return True
-        return False
 
     def _push(self, source: str, destination: str) -> str:
+        if self.client.project is None:
+            raise NotImplementedError(
```

Comment:
I think it might be worth using `PermissionError` instead of `NotImplementedError`, because we're never going to implement this for anonymous access 😄

Reply:
I briefly discussed that with @cdknorow above. From my point of view, it's not generally a `PermissionError`, because as far as I'm aware it is possible to configure a GCP bucket to allow anonymous writes. However, this is semantic pettifoggery and I'm happy to go with whatever you as the maintainer prefer.

How do you feel about a `NotSupportedError`?

Reply:
Ah, sorry, I forgot about the discussion above. Let's leave this as is for now. In the future it would be nice if modelstore could somehow check what permissions it has when doing anonymous access and then just work